src/Umbraco.Web/PropertyEditors/LabelPropertyEditor.cs

```diff
@@ -43,32 +41,94 @@ public override bool IsReadOnly
-            /// we should still show the pre-values stored for the data type.
+            /// Other than for the pre-value fields defined on this property editor, chuck all the values into one field so devs can see what is stored there.
+            /// We want this in case we've converted a legacy property editor over to a label as we should still show the pre-values stored for the data type.
             /// </summary>
             /// <param name="defaultPreVals"></param>
             /// <param name="persistedPreVals"></param>
             /// <returns></returns>
             public override IDictionary<string, object> ConvertDbToEditor(IDictionary<string, object> defaultPreVals, PreValueCollection persistedPreVals)
             {
                 var existing = base.ConvertDbToEditor(defaultPreVals, persistedPreVals);
-                //convert to a list, easier to enumerate on the editor
-                var asList = existing.Select(e => new KeyValuePair<string, object>(e.Key, e.Value)).ToList();
-                var result = new Dictionary<string, object> { { "values", asList } };
+
+                // Check for a saved value type.  If not found set to default string type.
+                var valueType = PropertyEditorValueTypes.StringType;
+                if (existing.ContainsKey(Constants.PropertyEditors.DataValueTypePreValueKey))
+                {
+                    valueType = (string)existing[Constants.PropertyEditors.DataValueTypePreValueKey];
+                }
+
+                // Convert any other values from a legacy property editor to a list, easier to enumerate on the editor.
+                // Make sure to exclude values defined on the label property editor itself.
+                var asList = existing
+                    .Select(e => new KeyValuePair<string, object>(e.Key, e.Value))
+                    .Where(e => e.Key != Constants.PropertyEditors.DataValueTypePreValueKey)
+                    .ToList();
+
+                var result = new Dictionary<string, object> { { Constants.PropertyEditors.DataValueTypePreValueKey, valueType } };
+                if (asList.Any())
+                {
+                    result.Add("values", asList);
+                }
+
                 return result;
             }
-        }
 
+            /// <summary>
+            /// When saving we want to avoid saving an empty "legacy property editor values" field if there are none.
+            /// </summary>
+            /// <param name="editorValue"></param>
+            /// <param name="currentValue"></param>
+            /// <returns></returns>
+            public override IDictionary<string, PreValue> ConvertEditorToDb(IDictionary<string, object> editorValue, PreValueCollection currentValue)
+            {
+                if (editorValue.ContainsKey(LegacyPropertyEditorValuesKey))

```

Comment:
Yes, this was a bit fiddly - but all stemmed from the fact that even though the label property editor could have pre-values (from a legacy type), they couldn't up to now be edited and saved through the UI.  Now that a "true" pre-value has been added, it can, which led to some odd behaviour.

Firstly there would be a pre-value record saved for legacy values even if there aren't any (the key would exist but with no value).  In that case I remove that pre-value so it's not saved (likely does no harm, but it's not necessary - we only need this legacy values pre-value record if there are any).

Secondly if there are legacy values, I found on each save the JSON structure containing them would get repeatedly nested (an outer JSON wrapper would be added each time).  So what I'm doing is if there are legacy pre-values, I'm converting what comes in "wrapped" like this...

```
 Value = {[
 {
     "Key": "values",
     "Value": {
		<legacy property editor values>
     }}
 ]}
```

...into this...

```
{
    <legacy property editor values>
}
```

... which takes it back to the originally saved value from when the label property type was created from the legacy one.